### PR TITLE
Send message functionality, Conversation functionality

### LINKED ISSUE #94: Create functionality to send message
- Create ChatMessage Entity
- Create ChatService
- Create Mapper
- Implement Controller method
- Partial Testing

#59

## Estimated time 
6h

### LINKED ISSUE #96: Create functionality to get all messages by user
- Add methods to  ChatService
- Create Mapper
- Implement Controller method
- Partial Testing

#59 

## Estimated time
6h

### LINKED ISSUE #97: Create functionality to get all conversations between 2 users
- Create Conversation Entity
 -Add methods to  ChatService
- Create Mapper
- Implement Controller method
- |Partial Testing

#59 

## Estimated time
6h

### LINKED ISSUE #101: Code adjustment for chat user stories
- Moving authorization of user implementation from Controllers to Services
- Introduction of new ChatMessagePostDTO DTO object
- New Mapping for  ChatMessagePostDTO DTO object
- Tests adjustments

#4 #5 

## Estimated time
3h

### PR DESCRIPTION
**Pull request by KseniaBel**
- Send message, get all messages, get all conversations by user functionality
#94

### CODE REVIEW

File: src/main/java/ch/uzh/ifi/hase/soprafs21/controller/MessageApiController.java [hunk 1]
```diff
@@ -12,13 +29,35 @@
...
+
     @Override
     public Optional<NativeWebRequest> getRequest() {
         return Optional.ofNullable(request);
     }
 
+    @Override
+    public ResponseEntity<Void> messagePost(@ApiParam(value = "A message that needs to be created" ,required=true )  @Valid @RequestBody ChatMessageDto chatMessageDto) throws Exception {
+        Long senderId = chatMessageDto.getSender().getId();
+        Long receiverId = chatMessageDto.getReceiver().getId();
```
Comment (bschne) [inline src/main/java/ch/uzh/ifi/hase/soprafs21/controller/MessageApiController.java]: Do we want to actually have the full user object in these DTOs? Seems a bit excessive, maybe we should include sender- and recipient ID in the ChatMessageDTO directly instead.

File: src/main/java/ch/uzh/ifi/hase/soprafs21/controller/MessageApiController.java [hunk 2]
```diff
@@ -12,13 +29,35 @@
...
+
+        String authenticatedUserEmail = (String) SecurityContextHolder.getContext().getAuthentication().getPrincipal();
+        User authenticatedUser = userService.getUserByEmail(authenticatedUserEmail).orElseThrow();
+        User sender = userService.getUserById(senderId);
+        if (!authenticatedUser.equals(sender)){
+            return ResponseEntity.status(HttpStatus.FORBIDDEN).build();
+        }
+
+        User receiver = userService.getUserById(receiverId);
+        chatService.createMessage(receiver, chatMessageDto.getMessage());
```
Comment (bschne) [inline src/main/java/ch/uzh/ifi/hase/soprafs21/controller/MessageApiController.java]: Maybe we should pass a full message entity to this service, including sender, receiver and message.

File: src/main/java/ch/uzh/ifi/hase/soprafs21/controller/UsersApiController.java [hunk 3]
```diff
@@ -54,6 +68,41 @@ public UsersApiController(NativeWebRequest request, UserService userService) {
...
 
+    @Override
+    public ResponseEntity<List<ChatMessageDto>> usersUserId1ConversationsUserId2Get(@ApiParam(value = "Numeric ID of the user1",required=true) @PathVariable("userId1") Long userId1, @ApiParam(value = "Numeric ID of the user2",required=true) @PathVariable("userId2") Long userId2) throws Exception {
+        if (!isRequesterAndAuthenticatedUserTheSame(userId1)){
+            return ResponseEntity.status(HttpStatus.FORBIDDEN).build();
+        }
+
+        User receiver = userService.getUserById(userId2);
+        try {
+            List<ChatMessage> messages = chatService.getAllMessages(receiver);
```
Comment (bschne) [inline src/main/java/ch/uzh/ifi/hase/soprafs21/controller/UsersApiController.java]: Again it might be cleaner to keep authentication-related logic in the controller layer and pass both users' IDs to the service here, instead of getting the currently logged-in user inside the service/repo.

File: src/main/java/ch/uzh/ifi/hase/soprafs21/controller/UsersApiController.java [hunk 4]
```diff
@@ -54,6 +68,41 @@ public UsersApiController(NativeWebRequest request, UserService userService) {
...
+    }
+
+    @Override
+    public ResponseEntity<List<ConversationDto>> usersUserIdConversationsGet(@ApiParam(value = "Numeric ID of the user",required=true) @PathVariable("userId") Long userId) throws Exception {
+        if (!isRequesterAndAuthenticatedUserTheSame(userId)){
+            return ResponseEntity.status(HttpStatus.FORBIDDEN).build();
+        }
+        User sender = userService.getUserById(userId);
+        try {
+            List<Conversation> conversations = chatService.getAllConversations();
```
Comment (bschne) [inline src/main/java/ch/uzh/ifi/hase/soprafs21/controller/UsersApiController.java]: Again, might be cleaner to pass the logged-in user's ID here instead of relying on auth in service/repo layers.

File: src/main/java/ch/uzh/ifi/hase/soprafs21/entity/ChatMessage.java [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+    @JoinColumn(name = "conversation_id", nullable = false)
+    @ManyToOne(fetch = FetchType.LAZY)
+    private Conversation conversation;
+
+
+    @PrePersist
+    private void updateTimeStamp() {
+        this.timeStamp = LocalDateTime.now();
+        unread = true;
+    }
```
Comment (bschne) [inline src/main/java/ch/uzh/ifi/hase/soprafs21/entity/ChatMessage.java:45]: Nice solution, didn't know about these hooks :-) 👍

File: src/main/java/ch/uzh/ifi/hase/soprafs21/service/ChatService.java [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+    private UserRepository userRepository;
+
+    public List<Conversation> getAllConversations(){
+        return conversationRepository.findByUser(getCurrentlyLoggedinUser().orElseThrow());
+    }
+
+    public List<ChatMessage> getAllMessages(User participant2) {
+        Conversation currentConversation = findConversationByParticipants(participant2).orElseThrow();
+        List<ChatMessage> allMessages = currentConversation.getMessages();
+        chatMessageRepository.setMessagesReadByConversation(currentConversation);
```
Comment (bschne) [inline src/main/java/ch/uzh/ifi/hase/soprafs21/service/ChatService.java:47]: IMO the frontend should make a second request to mark the convo as read when the page has loaded, since we don't want the GET-request to have side effects, e.g. also if the frontend crashes while loading the convo or something like that.

File: src/main/java/ch/uzh/ifi/hase/soprafs21/repository/ChatMessageRepository.java [hunk 7]
```diff
@@ -0,0 +1,24 @@
...
+import org.springframework.data.jpa.repository.Modifying;
+import org.springframework.data.jpa.repository.Query;
+
+import javax.transaction.Transactional;
+import java.util.List;
+
+public interface ChatMessageRepository extends JpaRepository<ChatMessage, Long> {
+    @Query("SELECT m FROM ChatMessage m WHERE m.sender = ?1 and m.timeStamp in " +
+            "(select max(m1.timeStamp) from ChatMessage m1 where m1.sender = ?1 GROUP BY m.receiver)")
+    List<ChatMessage> findAllLastMessagesBySender(User sender);
```
Comment (bschne) [inline src/main/java/ch/uzh/ifi/hase/soprafs21/repository/ChatMessageRepository.java]: Is this method used anywhere? Looks obsolete.

File: src/main/java/ch/uzh/ifi/hase/soprafs21/repository/ConversationRepository.java [hunk 8]
```diff
@@ -0,0 +1,19 @@
...
+import ch.uzh.ifi.hase.soprafs21.entity.User;
+import org.springframework.data.jpa.repository.JpaRepository;
+import org.springframework.data.jpa.repository.Query;
+
+import java.util.List;
+import java.util.Optional;
+
+public interface ConversationRepository extends JpaRepository<Conversation, Long> {
+    @Query("Select c from Conversation c where c.participant1 = ?1 or c.participant2 = ?1")
+    List<Conversation> findByUser(User user);
```
Comment (bschne) [inline src/main/java/ch/uzh/ifi/hase/soprafs21/repository/ConversationRepository.java:13]: I still don't think we need the conversation entity — couldn't we just query messages for `m.sender = ?1 or m.recipient = ?1` ? 

If we need to do that anyway, there's no point to having a separate conversation entity stored redundantly in our database. 

I think the conversation might make sense as a DTO so we have a neat object to pass user info and messages to the client, but not as a DB entity.

File: src/main/resources/application.yml [hunk 9]
```diff
@@ -30,11 +29,14 @@ spring:
 
   jpa:
     hibernate:
-      ddl-auto: update
+      ddl-auto: create
```
Comment (bschne) [inline src/main/resources/application.yml:32]: Why was this changed to `create`? Update should handle creation of new entities as well, I think.

File: src/test/java/ch/uzh/ifi/hase/soprafs21/controller/MessageApiControllerTest.java [hunk 10]
```diff
@@ -0,0 +1,82 @@
...
+    @BeforeEach
+    public void setup() {
+        MockitoAnnotations.openMocks(this);
+        when(authenticationMock.getPrincipal()).thenReturn("mark@twen.de");
+        when(securityContextMock.getAuthentication()).thenReturn(authenticationMock);
+        SecurityContextHolder.setContext(securityContextMock);
+
+        authenticatedUser = new User();
+        authenticatedUser.setId(1l);
+        authenticatedUser.setEmail("mark@twen.de");
```
Comment (bschne) [inline src/test/java/ch/uzh/ifi/hase/soprafs21/controller/MessageApiControllerTest.java:64]: image

File: src/test/java/ch/uzh/ifi/hase/soprafs21/controller/MessageApiControllerTest.java [hunk 11]
```diff
@@ -0,0 +1,82 @@
...
+        when(this.userServiceMock.getUserById(eq(2l))).thenReturn(receiverMock);
+        ChatMessageDto message = new ChatMessageDto();
+        message.setMessage("Hello");
+        message.setReceiver(receiver);
+        message.setSender(sender);
+
+        ResponseEntity<Void> responseEntity = messageApiController.messagePost(message);
+        verify(chatServiceMock).createMessage(eq(receiverMock), eq(message.getMessage()));
+        assertEquals(HttpStatus.CREATED, responseEntity.getStatusCode());
+    }
```
Comment (bschne) [inline src/test/java/ch/uzh/ifi/hase/soprafs21/controller/MessageApiControllerTest.java]: We should probably add at least the 301 and 404 cases as well.

File: src/test/java/ch/uzh/ifi/hase/soprafs21/service/ChatServiceTest.java [hunk 12]
```diff
@@ -0,0 +1,108 @@
...
+        List<Conversation> conversations = chatService.getAllConversations();
+        assertEquals(2, conversations.size());
+        assertEquals(5, conversations.get(0).getMessages().size());
+        LocalDateTime currentDateTime = LocalDateTime.now();
+
+        for(ChatMessage chatMessage: conversations.get(0).getMessages()) {
+            assertTrue(currentDateTime.isAfter(chatMessage.getTimeStamp()));
+            currentDateTime = chatMessage.getTimeStamp();
+        }
+    }
```
Comment (bschne) [inline src/test/java/ch/uzh/ifi/hase/soprafs21/service/ChatServiceTest.java]: Is this reliant on the initializer SQL file above? If so, is that file only used in the test context, or also when running our application for the dev/prod environments? If it's the latter, we should ideally set up that test data for tests only so it's isolated and won't change, but it's probably fine for now.

**Review by bschne — CHANGES_REQUESTED** [review]
Overall good work, there's a few details I marked and two bigger points I think we should address.

**Conversation Entity**
I think `Conversation` has justification as a DTO to neatly pass user- and message info to the client in one request, but not as a separate entity in the DB since this introduces redundancy unnecessarily. IMO we should be able to query the messages by doing `m.sender = ?1 or m.recipient = ?1`, which we need to do with conversations anyway since we don't know who is user1 and who is user2 in a conversation as this depends on who messaged first. Caveat: I might be missing something here since I didn't implement it — if so just let me know.

**Relying on Auth in Services & Repos**
As this code stands, the services and repositories make retrieve the currently logged-in user through helper methods where needed and use it to e.g. determine who is the sender of a message. I think we should keep this out and pass these IDs in as parameters wherever we call such a service method from the controller, with the service then passing it as a parameter to the repository again. This would be cleaner and reduce the reliance of repos & services on external state (except the database ofc., but that is their purpose) and thus easier to reason about and unit test.

File: src/main/java/ch/uzh/ifi/hase/soprafs21/repository/ConversationRepository.java (hunk 8 above)
Comment (KseniaBel) [reply src/main/java/ch/uzh/ifi/hase/soprafs21/repository/ConversationRepository.java:13]: This DB normalization is done to present in useful way the data through REST API. Without Conversation JPQL query would be too complicated to implement and we will need to group/process messages  programmatically. I tried to write JPQL without Conversation and I failed. (your query proposal does not include group by sender/recipient and eventually sorted messages).

File: src/main/resources/application.yml (hunk 9 above)
Comment (KseniaBel) [reply src/main/resources/application.yml:32]: This was not a case for the integration tests, the sequence of the primary keys was not created.

File: src/test/java/ch/uzh/ifi/hase/soprafs21/controller/MessageApiControllerTest.java (hunk 11 above)
Comment (KseniaBel) [reply src/test/java/ch/uzh/ifi/hase/soprafs21/controller/MessageApiControllerTest.java]: Did you mean 401?

Comment (bschne) [reply src/test/java/ch/uzh/ifi/hase/soprafs21/controller/MessageApiControllerTest.java]: 401 and 404, yes, sorry

File: src/test/java/ch/uzh/ifi/hase/soprafs21/service/ChatServiceTest.java (hunk 12 above)
Comment (KseniaBel) [reply src/test/java/ch/uzh/ifi/hase/soprafs21/service/ChatServiceTest.java]: Used only for the integration test. What do you mean by "set up that test data for tests only"?

File: src/main/java/ch/uzh/ifi/hase/soprafs21/controller/UsersApiController.java (hunk 3 above)
Comment (KseniaBel) [reply src/main/java/ch/uzh/ifi/hase/soprafs21/controller/UsersApiController.java]: This is not an authentication logic, but authorization check. Since you mentioned this, it should be done in Service layer and I will move it there.

File: src/main/java/ch/uzh/ifi/hase/soprafs21/controller/MessageApiController.java (hunk 2 above)
Comment (KseniaBel) [reply src/main/java/ch/uzh/ifi/hase/soprafs21/controller/MessageApiController.java]: Why do you think it matters? I pass them like this because I need just the receiver and the message text and no other attributes are needed.

File: src/main/java/ch/uzh/ifi/hase/soprafs21/controller/MessageApiController.java (hunk 1 above)
Comment (KseniaBel) [reply src/main/java/ch/uzh/ifi/hase/soprafs21/controller/MessageApiController.java]: They are not full User objects but UserOverviewDto (which is already simplified ). We could create a separate obj for a messege POST as well.

**Comment by KseniaBel** [comment]
> Overall good work, there's a few details I marked and two bigger points I think we should address.
> 
> **Conversation Entity**
> I think `Conversation` has justification as a DTO to neatly pass user- and message info to the client in one request, but not as a separate entity in the DB since this introduces redundancy unnecessarily. IMO we should be able to query the messages by doing `m.sender = ?1 or m.recipient = ?1`, which we need to do with conversations anyway since we don't know who is user1 and who is user2 in a conversation as this depends on who messaged first. Caveat: I might be missing something here since I didn't implement it — if so just let me know.
> 
> **Relying on Auth in Services & Repos**
> As this code stands, the services and repositories make retrieve the currently logged-in user through helper methods where needed and use it to e.g. determine who is the sender of a message. I think we should keep this out and pass these IDs in as parameters wherever we call such a service method from the controller, with the service then passing it as a parameter to the repository again. This would be cleaner and reduce the reliance of repos & services on external state (except the database ofc., but that is their purpose) and thus easier to reason about and unit test.

1. Regarding the conversation entity: https://github.com/sopra-fs21-group-02/server/pull/95#discussion_r607980441
2. Authentication/identity verification should be performed by OAuth2 client/Spring security on the REST layer, while authorization checks such as "if user is allowed to do some action" has to be done on service layer (I will move my code from Controller to Service).

File: src/main/java/ch/uzh/ifi/hase/soprafs21/service/ChatService.java (hunk 6 above)
Comment (KseniaBel) [reply src/main/java/ch/uzh/ifi/hase/soprafs21/service/ChatService.java:47]: It is valid point. We could make a follow up issue to correct it if we have the time.

File: src/main/java/ch/uzh/ifi/hase/soprafs21/repository/ChatMessageRepository.java (hunk 7 above)
Comment (KseniaBel) [reply src/main/java/ch/uzh/ifi/hase/soprafs21/repository/ChatMessageRepository.java]: To be deleted. Thanks

File: src/main/java/ch/uzh/ifi/hase/soprafs21/controller/UsersApiController.java (hunk 3 above)
Comment (bschne) [reply src/main/java/ch/uzh/ifi/hase/soprafs21/controller/UsersApiController.java]: Yes, sorry about mixing up the terms — I was thinking about the fact that the service layer accesses the current user, instead of receiving a userId from the controller. I was aiming to have only the controllers access the current user from the context, and then pass its ID to the lower layers.

That being said you are right, it's probably better to have this in the service layer.

File: src/main/java/ch/uzh/ifi/hase/soprafs21/controller/MessageApiController.java (hunk 2 above)
Comment (bschne) [reply src/main/java/ch/uzh/ifi/hase/soprafs21/controller/MessageApiController.java]: Was mostly related to what I had in mind re. not fetching the current user from the context in services, but that point is obsolete now :)

File: src/main/java/ch/uzh/ifi/hase/soprafs21/controller/MessageApiController.java (hunk 1 above)
Comment (bschne) [reply src/main/java/ch/uzh/ifi/hase/soprafs21/controller/MessageApiController.java]: We will not need all the info in UserOverviewDto with each message when we get the full conversation, and even less so when we post a new message.

I think this should only contain the message and sender/recipient IDs. But we can also adjust it as needed when we implement the corresponding frontend parts and leave it like this for now.

File: src/main/java/ch/uzh/ifi/hase/soprafs21/repository/ConversationRepository.java (hunk 8 above)
Comment (bschne) [reply src/main/java/ch/uzh/ifi/hase/soprafs21/repository/ConversationRepository.java:13]: I see the issue now with the query we need for the "inbox" part, sorry. I missed the grouping aspect there before. 

Not worth the hassle, let's keep it like this 👍

File: src/test/java/ch/uzh/ifi/hase/soprafs21/service/ChatServiceTest.java (hunk 12 above)
Comment (bschne) [reply src/test/java/ch/uzh/ifi/hase/soprafs21/service/ChatServiceTest.java]: I was not sure if the data in the sql file a) gets loaded into the database after starting the application (regardless of whether it's running tests, started as a dev server, or started in production) or b) is only used when running tests. If it's a), I was saying that it might be better to not rely on data defined outside the tests in our test cases, i.e. that the tests should create all the data they base their assertions on themselves in a setup method or similar, not that our application *always* populates the DB (regardless of whether it's run in production, for the tests, ...) with some data that is then used for testing.

File: src/main/resources/application.yml (hunk 9 above)
Comment (bschne) [reply src/main/resources/application.yml:32]: OK, let's leave it in that case. From what I've read I think this leads to all previous data being dropped on application start. We still have it on 'auto' in application-production.yml though, so it shouldn't be a big issue.

File: src/main/java/ch/uzh/ifi/hase/soprafs21/controller/MessageApiController.java (hunk 1 above)
Comment (KseniaBel) [reply src/main/java/ch/uzh/ifi/hase/soprafs21/controller/MessageApiController.java]: I see your point that we do not need to have some of the fields of receiver/sender  when we get a message in our current implementation. But I did it this way so that the code is more reusable. In case we decide that we need more info of the message sender, for example. 

For the POST I will create the DTO with receiver and sender IDs only.

File: src/main/resources/application.yml (hunk 9 above)
Comment (KseniaBel) [reply src/main/resources/application.yml:32]: url: jdbc:h2:mem:soprafsdb;MODE=PostgreSQL;IGNORECASE=TRUE;INIT=CREATE SCHEMA IF NOT EXISTS soprafs21

As far as I can see we have in-memory DB, so the previous data will be dropped anyway. Or do I miss something?

Comment (bschne) [reply src/main/resources/application.yml:32]: I'm being stupid today, I had the `mem:soprafsdb` part changed to use a file instead of in-memory at some point when I was setting things up but then changed it again b/c I had some issue before pushing it and forgot about it, sorry.

File: src/test/java/ch/uzh/ifi/hase/soprafs21/service/ChatServiceTest.java (hunk 12 above)
Comment (KseniaBel) [reply src/test/java/ch/uzh/ifi/hase/soprafs21/service/ChatServiceTest.java]: I see. It is not "data.sql" it is "data_init.sql", so it won't be loaded automatically on starting the application. I can rename it to something else not to create confusion.

Comment (bschne) [reply src/test/java/ch/uzh/ifi/hase/soprafs21/service/ChatServiceTest.java]: The confusion was all mine, you can leave it like this

File: src/main/resources/application.yml (hunk 9 above)
Comment (KseniaBel) [reply src/main/resources/application.yml:32]: If you change it to file-based DB, please keep it in build folder for now.

**Review by bschne — APPROVED** [review]
👍
